File: frontend/src/types/project.ts
import type { Job } from "./job";

export type Project = {
  id: string;
  name: string;
  description: string | null;
  status: "no_model" | "processing" | "ready" | "failed";
  image_upload_id: string | null;
  model_asset_id: string | null;
  last_job_id: string | null;
  edit_version: number;
  created_at: string;
  updated_at: string;
};

export type ModelStats = {
  vertices: number;
  faces: number;
  edges: number;
  triangles: number;
  bounds?: [number, number, number][] | null;
  has_texture: boolean;
  texture_size?: number | null;
  watertight?: boolean;
};

export type ModelSummary = {
  filename: string;
  size_bytes: number;
  mime_type: string;
  stats: ModelStats | null;
  created_at: string;
  download_glb_url: string | null;
  download_gltf_url: string | null;
  download_obj_url?: string | null;
  preview_url?: string | null;
};

export type ProjectDetail = Project & {
  image_url?: string | null;
  model_summary?: ModelSummary | null;
  last_job?: Job | null;
};

export type ProjectCreateRequest = {
  name: string;
  description?: string;
};

export type ProjectUpdateRequest = {
  name?: string;
  description?: string | null;
};

export type ProjectFilter = {
  query?: string;
  status?: string;
  page?: number;
  page_size?: number;
  sort_by?: string;
  sort_dir?: "asc" | "desc";
};

File: backend/app/schemas/project.py
"""Project schemas."""
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import Field

from app.schemas.common import ApiModel


class ProjectCreateRequest(ApiModel):
    name: str = Field(min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=500)


class ProjectRenameRequest(ApiModel):
    name: str = Field(min_length=1, max_length=200)


class ProjectUpdateRequest(ApiModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=500)


class ProjectPublic(ApiModel):
    id: str
    name: str
    description: Optional[str] = None
    status: str
    image_upload_id: Optional[str] = None
    model_asset_id: Optional[str] = None
    last_job_id: Optional[str] = None
    edit_version: int = 0
    created_at: datetime
    updated_at: datetime


class ProjectDetail(ProjectPublic):
    image_url: Optional[str] = None
    model_summary: Optional["ModelSummary"] = None
    last_job: Optional["JobPublic"] = None


class ProjectFilter(ApiModel):
    query: Optional[str] = None
    status: Optional[str] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=12, ge=1, le=100)
    sort_by: str = "created_at"
    sort_dir: str = "desc"


from app.schemas.job import JobPublic  # noqa: E402
from app.schemas.model import ModelSummary  # noqa: E402

ProjectDetail.model_rebuild()
